omnigibson/controllers/osc_controller.py

```diff
@@ -0,0 +1,535 @@
+                    command_output_limits = [
+                        [-1.0, -1.0, -1.0, -np.pi, -np.pi, -np.pi],
+                        [1.0, 1.0, 1.0, np.pi, np.pi, np.pi],
+                    ]
+                else:
+                    command_output_limits[0][3:] = -np.pi
+                    command_output_limits[1][3:] = np.pi
+
+        is_input_limits_numeric = not (command_input_limits is None or isinstance(command_input_limits, str))
+        is_output_limits_numeric = not (command_output_limits is None or isinstance(command_output_limits, str))
+        command_input_limits = [nums2array(lim, dim=6, dtype=np.float32) for lim in command_input_limits] if is_input_limits_numeric else command_input_limits
+        command_output_limits = [nums2array(lim, dim=6, dtype=np.float32) for lim in command_output_limits] if is_output_limits_numeric else command_output_limits
+
+        # Modify input / output scaling based on whether we expect gains to be part of the action space
+        self._command_dim = OSC_MODE_COMMAND_DIMS[self.mode]
+        for variable_gain, gain_limits, dim in zip(
+                (self.variable_kp, self.variable_damping_ratio, self.variable_kp_null),
+                (self.kp_limits, self.damping_ratio_limits, self.kp_null_limits),
+                (6, 6, control_dim),
+        ):
+            if variable_gain:
+                # Add this to input / output limits
+                if is_input_limits_numeric:
+                    command_input_limits = [np.concatenate([lim, nums2array(nums=val, dim=dim, dtype=np.float32)]) for lim, val in zip(command_input_limits, (-1, 1))]
+                if is_output_limits_numeric:
+                    command_output_limits = [np.concatenate([lim, nums2array(nums=val, dim=dim, dtype=np.float32)]) for lim, val in zip(command_output_limits, gain_limits)]
+                # Update command dim
+                self._command_dim += dim
+
+        # Other values
+        self.decouple_pos_ori = decouple_pos_ori
+        self.workspace_pose_limiter = workspace_pose_limiter
+        self.task_name = task_name
+        self.default_joint_pos = default_joint_pos[dof_idx].astype(np.float32)
+
+        # Other variables that will be filled in at runtime
+        self.goal_pos = None
+        self.goal_ori_mat = None
+        self._quat_target = None
+
+        # Run super init
+        super().__init__(
+            control_freq=control_freq,
+            control_limits=control_limits,
+            dof_idx=dof_idx,
+            command_input_limits=command_input_limits,
+            command_output_limits=command_output_limits,
+        )
+
+    def reset(self):
+        # Clear internal variables
+        self.goal_pos = None
+        self.goal_ori_mat = None
+        self._quat_target = None
+        self._clear_variable_gains()
+
+    @property
+    def state_size(self):
+        # Add 4 for internal quat target
+        # TODO: Store goal pose, goal ori mat, gains, etc.
```

Comment:
is this TODO going to be completed within this PR?

Reply:
sorry, will be done in next one